### PR TITLE
[Quant] Add fused LinearLeakyReLU module for onednn backend

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89188
* #88923
* #88879
* #88668
* #88665
* __->__ #88661

**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `QLinearLeakyReLU` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by jerryzh168** [comment]
should we add a test in https://github.com/pytorch/pytorch/blob/master/test/quantization/eager/test_quantize_eager_ptq.py#L76

**Comment by Xia-Weiwen** [comment]
> should we add a test in https://github.com/pytorch/pytorch/blob/master/test/quantization/eager/test_quantize_eager_ptq.py#L76

Thanks for the suggestion. But I did not see tests for fused module here.
Shall we add it here?
https://github.com/pytorch/pytorch/blob/a7420d2ccb62d005f2e1853cfef8d25eb7748a90/test/quantization/core/test_quantized_module.py#L49
If so, do you suggest adding a new test implementation for `LinearLeakyReLU` or modify `_test_linear_api_impl` below so that it supports more fusions?
https://github.com/pytorch/pytorch/blob/a7420d2ccb62d005f2e1853cfef8d25eb7748a90/test/quantization/core/test_quantized_module.py#L83

**Review by jgong5 — APPROVED** [review]
LGTM on the changes. Consider to add a test as Jerry suggested.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 I have added a test here: test/quantization/core/test_quantized_module.py. Please take a look.

File: test/quantization/core/test_quantized_module.py [hunk 1]
```diff
@@ -72,22 +73,28 @@ def test_linear_api(self):
...
-            False: nnq.Linear,
+            'none': nnq.Linear,
+            'relu': nniq.LinearReLU,
+            'leaky_relu': nniq.LinearLeakyReLU,
+        }
+        q_module_name_map = {
+            'none': 'QuantizedLinear',
+            'relu': 'QuantizedLinearReLU',
+            'leaky_relu': 'QuantizedLinearLeakyReLU',
         }
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_module.py:98]: I think we should probably pass these from argument, it looks a bit weird that we hardcode these in this helper function

Comment (jerryzh168) [reply test/quantization/core/test_quantized_module.py:98]: also for naming, I think we can rename `test_linear_api` to `test_linear` as well, to make it clearer

Comment (jerryzh168) [reply test/quantization/core/test_quantized_module.py:98]: and we can split `test_linear` to `test_linear` and `test_linear_relu` to make sure each test only tests one thing

**Comment by jerryzh168** [comment]
please add the summary and test plan as suggested in previous PR as well. Also are we planning to use this in eager mode as well or just fx graph mode? and how urgent is this change? can you wait for PyTorch 2.0 where we integration quantization in PyTorch 2.0 export path and these changes may not be needed?

File: test/quantization/core/test_quantized_module.py (hunk 1 above)
Comment (Xia-Weiwen) [reply test/quantization/core/test_quantized_module.py:98]: OK I will do that.

File: torch/ao/nn/intrinsic/quantized/modules/linear_relu.py [hunk 2]
```diff
@@ -39,3 +40,68 @@ def from_float(cls, mod):
...
+        qlinear.set_weight_bias(qweight, mod.bias)
+        qlinear.scale = float(act_scale)
+        qlinear.zero_point = int(act_zp)
+        return qlinear
+
+    @classmethod
+    def from_reference(cls, ref_mod, output_scale, output_zero_point):
+        linear = ref_mod[0]
+        leaky_relu = ref_mod[1]
+        qlinear = cls(
```
Comment (jerryzh168) [inline torch/ao/nn/intrinsic/quantized/modules/linear_relu.py:98]: nit: qlinear -> qlinear_leakyrelu to make it clearer

**Comment by Xia-Weiwen** [comment]
> please add the summary and test plan as suggested in previous PR as well.

Ok. I will add them later.

> Also are we planning to use this in eager mode as well or just fx graph mode? 

It's just for FX mode

> and how urgent is this change? can you wait for PyTorch 2.0 where we integration quantization in PyTorch 2.0 export path and these changes may not be needed?

Not that urgent but it's nice to land ASAP. I wander how quantization path will be like in PyTorch 2.0. Do you mean that there will be a different mechanism for fusion, so these PRs are not needed? If so, we would like to align with 2.0 to support more fusion patterns. How to do that? Thanks.

**Comment by jerryzh168** [comment]
> > Also are we planning to use this in eager mode as well or just fx graph mode?
> 
> It's just for FX mode
> 

In that case probably better to postpone a little bit if it's possible (e.g. next year, when we have PyTorch 2.0 export open sourced)

> Not that urgent but it's nice to land ASAP. I wander how quantization path will be like in PyTorch 2.0. Do you mean that there will be a different mechanism for fusion, so these PRs are not needed? If so, we would like to align with 2.0 to support more fusion patterns. How to do that? Thanks.

please stay tuned to upcoming PyTorch Developer Conference: https://pytorch.org/blog/announcing-pytorch-conference-2022/, we'll have a poster for quantization as well. but in short, yeah these PRs won't be needed in that stack, we'll be working with aten operators and backend developers will only need to write 1. BackendConfig 2. quantized operators or delegated operators 3. lowering passes from reference quantized pattern to the quantized operators/delegated operators, especially for post training quantization

**Comment by Xia-Weiwen** [comment]
> > > Also are we planning to use this in eager mode as well or just fx graph mode?
> > 
> > 
> > It's just for FX mode
> 
> In that case probably better to postpone a little bit if it's possible (e.g. next year, when we have PyTorch 2.0 export open sourced)
> 
> > Not that urgent but it's nice to land ASAP. I wander how quantization path will be like in PyTorch 2.0. Do you mean that there will be a different mechanism for fusion, so these PRs are not needed? If so, we would like to align with 2.0 to support more fusion patterns. How to do that? Thanks.
> 
> please stay tuned to upcoming PyTorch Developer Conference: https://pytorch.org/blog/announcing-pytorch-conference-2022/, we'll have a poster for quantization as well. but in short, yeah this PR won't be needed in that stack, we'll be working with aten operators and backend developers will only need to write 1. BackendConfig 2. quantized operators or delegated operators 3. lowering passes from reference quantized pattern to the quantized operators/delegated operators, especially for post training quantization

Good to know that. Thanks. Also cc. @jgong5 @leslie-fang-intel

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. I'd like to ask a little more about PyTorch 2.0 quantization fusion. As you mentioned,
> backend developers will only need to write 1. BackendConfig 2. quantized operators or delegated operators 3. lowering passes from reference quantized pattern to the quantized operators/delegated operators

It looks however much the same as the current path, where we also need to prepare (1) backend config, (2) quantized fused ops and modules, and (3) lowering methods (fusers and mappings). So, which part will be different? Or to be more specific, which of the four PRs to add `linear - leaky_relu` fusion will not be needed in PyTorch 2.0? Thanks.

**Comment by jerryzh168** [comment]
> Hi @jerryzh168. I'd like to ask a little more about PyTorch 2.0 quantization fusion. As you mentioned,
> 
> > backend developers will only need to write 1. BackendConfig 2. quantized operators or delegated operators 3. lowering passes from reference quantized pattern to the quantized operators/delegated operators
> 
> It looks however much the same as the current path, where we also need to prepare (1) backend config, (2) quantized fused ops and modules, and (3) lowering methods (fusers and mappings). So, which part will be different? Or to be more specific, which of the four PRs to add `linear - leaky_relu` fusion will not be needed in PyTorch 2.0? Thanks.

There is one big change in the PT 2.0 export which is quantization will be working with aten operators instead of torch ops/modules, so we won't see the modules to start with, and we won't need fused modules etc.

But we can land this PR since that is still at least a few months away

**Comment by Xia-Weiwen** [comment]
> > Hi @jerryzh168. I'd like to ask a little more about PyTorch 2.0 quantization fusion. As you mentioned,
> > > backend developers will only need to write 1. BackendConfig 2. quantized operators or delegated operators 3. lowering passes from reference quantized pattern to the quantized operators/delegated operators
> > 
> > 
> > It looks however much the same as the current path, where we also need to prepare (1) backend config, (2) quantized fused ops and modules, and (3) lowering methods (fusers and mappings). So, which part will be different? Or to be more specific, which of the four PRs to add `linear - leaky_relu` fusion will not be needed in PyTorch 2.0? Thanks.
> 
> There is one big change in the PT 2.0 export which is quantization will be working with aten operators instead of torch ops/modules, so we won't see the modules to start with, and we won't need fused modules etc.
> 
> But we can land this PR since that is still at least a few months away

Thanks for the clarification. I will continue working on them.

File: test/quantization/core/test_quantized_module.py (hunk 1 above)
Comment (Xia-Weiwen) [reply test/quantization/core/test_quantized_module.py]: I have made these as arguments of the helper function and split `test_qlinear` to two tests for `qlinear` and `qlinear_relu` respectively.

File: torch/ao/nn/intrinsic/quantized/modules/linear_relu.py (hunk 2 above)
Comment (Xia-Weiwen) [reply torch/ao/nn/intrinsic/quantized/modules/linear_relu.py]: It's fixed

**Comment by Xia-Weiwen** [comment]
> please add the summary and test plan as suggested in previous PR as well.

They are added

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. Do you have more comments? Thanks!

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. Is it ok to land? Thanks

**Comment by jerryzh168** [comment]
yeah, looks good, please feel free to land

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168  You did not approve this. Could you approve it? Thanks!

**Review by jerryzh168 — APPROVED** [review]

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
sorry, missed this before, please revert changes to `torch/nn/intrinsic/` folder, we want to deprecate this namespace and we should not add new things to it

**Comment by Xia-Weiwen** [comment]
> sorry, missed this before, please revert changes to `torch/nn/intrinsic/` folder, we want to deprecate this namespace and we should not add new things to it

I have removed them.

**Review by jerryzh168 — APPROVED** [review]

**Pull request closed without merging**